python/tvm/hybrid/parser.py

```diff
@@ -0,0 +1,754 @@
+            Column offset of error line
+        """
+
+        if lineno is None:
+            lineno = self.current_lineno
+        if col_offset is None:
+            col_offset = self.current_col_offset
+        raise HybridParserError(self.wrap_line_col(message, lineno, col_offset))
+
+    def get_type_name(self, vtype):
+        if isinstance(vtype, ast.Attribute) \
+                and isinstance(vtype.value, ast.Name) and vtype.value.id == 'ty':
+            return vtype.attr
+        self.report_error("invalid type annotation")
+
+    def get_body(self):
+        body = []
+        while len(self.scope_emitter.node_stack[-1]) > 0:
+            res = self.visit(self.scope_emitter.node_stack[-1].pop())
+            if res is not None:
+                body.append(res)
+        return tvm.tir.SeqStmt(body) if len(body) > 1 else body[0]
+
+    def parse_type(self, vtype):
+        """ Parse type annotation AST into Type object """
+        if isinstance(vtype, ast.NameConstant) and vtype.value is None:
+            return tvm.ir.TupleType([])
+        elif isinstance(vtype, ast.Attribute):
+            return tvm.ir.PrimType(self.get_type_name(vtype))
+        elif isinstance(vtype, ast.Subscript) and isinstance(vtype.slice, ast.Index):
+            type_name = self.get_type_name(vtype.value)
+            if isinstance(vtype.slice.value, ast.Tuple):
+                args = [self.parse_type(element) for element in vtype.slice.value.elts]
+            else:
+                args = [self.parse_type(vtype.slice.value)]
+            if type_name == "Ptr":
+                return tvm.ir.PointerType(*args)
+            elif type_name == "Tuple":
+                return tvm.ir.TupleType(args)
+
+        self.report_error("invalid type annotation")
+
+    def generic_visit(self, node):
+        """ Override method in ast.NodeVisitor.
+        To directly filter out invalidate type of stmt.
+        """
+
+        self.report_error(type(node).__name__ + " stmt is not supported now")
+
+    def visit_Module(self, node):
+        """ Module visitor
+        AST abstract grammar:
+            Module(stmt* body, type_ignore* type_ignore)
+        By now we support two format of hybrid script shown below.
+
+        Example
+        -------
+        1. Generate a Function(If the code is printed, then it may bring meta)
+        .. code-block:: python
+
+            import tvm
+
+            @tvm.hybrid.script
+            def A(...):
+                ...
+
+            # call hybrid parser when call this function, get a Function
+            func = A
+
+        2. Generate a Module
+        .. code-block:: python
+
+            import tvm
+
+            @tvm.hybrid.script
+            class MyMod():
+               def A(...):
+                  ...
+
+               def B(...):
+                   ...
+
+                __tvm_meta__ = ...
+
+            # call hybrid parser during construction, get a Module
+            mod = MyMod
+        """
```

Comment:
shall we update the comment? is bracket needed?

Reply:
updated